apps/ejabberd/src/mongoose_commands.erl

```diff
@@ -0,0 +1,567 @@
+%% permissions required to run this command, defaults to [admin]
+%%      result :: argspec()
+%% Type spec of return value of the function to call; execute/3 eventually returns {ok, result}
+%%      identifiers :: [atom()] (optional, required in 'update' commands)
+%%
+%% You can ignore return value of the target func by specifying return value as {result, ok}. The
+%% execute/3 will then always return just 'ok' (or error).
+%%
+%% If action is 'update' then it MUST specify which args are to be used as identifiers of object to update.
+%% It has no effect on how the engine does its job, but may be used by client code to enforce proper
+%% structure of request. (this is bad programming practice but we didn't have a better idea, we had to
+%% solve it for REST API)
+%%
+%% Commands are registered here upon the module's initialisation
+%% (the module has to explicitly call mongoose_commands:register_commands/1
+%% func, it doesn't happen automagically), also should be unregistered when module
+%% terminates.
+%%
+%% Commands are executed by calling mongoose_commands:execute/3 method. This
+%% can return:
+%% {ok, Result}
+%% {error, denied, Msg} if user has no permission
+%% {error, not_implemented, Msg}
+%% {error, type_error, Msg} if either arguments or return value does not match
+%% {error, internal, Msg} if an exception was caught
+%%
+%% ==== Type check ====
+%%
+%% A command's definition includes specification of it arguments; when
+%% it is called, arguments are check for compatibility. Examples of specs
+%% and compliant arguments:
+%%
+%% a single type spec
+%% integer                          2
+%% binary                           <<"zzz">>
+%% atom                             brrr
+%% a list of arbitrary length, of a given type
+%% [integer]                        []
+%% [integer]                        [1]
+%% [integer]                        [1,2,3,4]
+%% a list of anything
+%% []
+%% a named argument (name is only for clarity)
+%% {msg, binary}                    <<"zzz">>
+%% a tuple of args
+%% {integer, binary, float}         {1, <<"2">>, 3.0}
+%% a tuple of named args
+%% {{x, integer}, {y, binary}}      {1, <<"bbb">>}
+%%
+%% Arg specification is used at call-time for control, and also for introspection
+%% (see list/1, list/2, mongoose_commands:get_command/2 and args/1)
+%%
+%% Return value is also specified, and this is a bit tricky: command definition
+%% contains spec of return value - what the target func returns should comply to it.
+%% The registry, namely execute/3, returns a tuple {ok, ValueReturnedByTheFunction}
+%% If return value is defined as 'ok' then whatever target func returns is ignored.
+%% This is mostly to make a distinction between 'create' actions which actually create something
+%% and return its identifier and those 'lame creators' which cause some action to be done and
+%% something written to dbase (exemplum: sending a message), but there is no accessible resource.
```

Comment:
I think we can remove this "lame creators". F.e if we assume, that the server will generate an `id` for message (and we, IMO, want to have ids for messages), the `id` may be returned.